crates/edr_solidity/src/contracts_identifier/radix_tree.rs

```diff
@@ -0,0 +1,416 @@
+
+                    // the current node now points to the new node
+                    self.child_nodes.insert(b, node);
+
+                    return;
+                }
+
+                // The content includes some part of the word, but not all of it
+                // insert a new in-between node between current node and it's child, that
+                // will have children for the old child and a new node for the given word.
+                let mut middle_node = RadixNode::new(
+                    word[..prefix_length].to_vec(),
+                    false,
+                    self.bytes_matched_before + self.content.len(),
+                );
+
+                // next_node should come after middle_node and its content and bytes_matched_before need to be adapted
+                next_node.content = next_node.content[prefix_length..].to_vec();
+                next_node.bytes_matched_before +=
+                    middle_node.bytes_matched_before + middle_node.content.len();
+                middle_node
+                    .child_nodes
+                    .insert(next_node.content[0], next_node);
+
+                // create a new node for the word
+                let new_node = RadixNode::new(
+                    word[prefix_length..].to_vec(),
+                    true,
+                    middle_node.bytes_matched_before + middle_node.content.len(),
+                );
+                middle_node
+                    .child_nodes
+                    .insert(word[prefix_length], new_node);
+
+                // set the middle_node as current_node's child
+                self.child_nodes.insert(b, middle_node);
+            }
+        }
+    }
+
+    /**
+     * Returns a tuple containing:
+     * - a boolean indicating if the word was matched exactly
+     * - the number of bytes matched
+     * - the node that matched the word
+     * If the word is not matched exactly, the node will be the one that matched the longest prefix.
+     */
+    fn get_max_match(&self, word: &[u8]) -> (bool, usize, &RadixNode) {
+        let prefix_length = get_shared_prefix_length(word, &self.content);
+
+        let matched = prefix_length + self.bytes_matched_before;
+
+        let entire_word_matched = prefix_length == word.len();
+        let entire_content_matched = prefix_length == self.content.len();
+
+        if entire_word_matched {
+            if entire_content_matched {
+                return (self.is_present, matched, &self);
+            }
+
+            return (false, matched, &self);
+        }
+
+        if !entire_content_matched {
+            return (false, matched, &self);
+        }
+
+        let next_node = self.child_nodes.get(&word[prefix_length]);
+
+        match next_node {
+            None => (false, matched, &self),
+            Some(next_node) => next_node.get_max_match(&word[prefix_length..]),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct RadixTree {
```

Comment:
Is something about this implementation of a radix tree different compared to existing crates; like https://crates.io/crates/radix-tree?

Reply:
To be honest, I didn't even consider using a crate for this. But I think having our own implementation makes sense, since this is a relatively simple data structure and a very core part of the tracing logic, so having flexibility seems important. Anyway, I'm not against using an external crate if you think it's a better idea, but in that case we should:

- Make sure that it does the same that our implementation does
- Implement the `getDescendantSuffixes` for it (it doesn't seem to have an equivalent function)

I'd say we complete the implementation of `ContractsIdentifier` with this implementation and then decide if we want to swap the radix tree with an external crate.